akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +351,120 @@ final class CoordinatedShutdown private[akka] (
+          }
+          if (p.isCompleted) {
+            Future.successful(Done)
+          } else {
+            val f = try {
+              task.apply().recover {
+                case NonFatal(exc) if recoverEnabled =>
+                  log.warning("Task [{}] failed in phase [{}]: {}", name, phaseName, exc.getMessage)
+                  Done
+              }
+            } catch { // in case task.apply() throws
+              case NonFatal(exc) if recoverEnabled =>
+                log.warning(
+                  "Task [{}] in phase [{}] threw an exception before its future could be constructed: {}",
+                  name,
+                  phaseName,
+                  exc.getMessage)
+                Future.successful(Done)
+              case NonFatal(exc) =>
+                Future.failed(exc)
+            }
+            p.completeWith(f)
+            f
+          }
+        }
+
+        override def cancel(): Boolean = {
+          val cancelled = p.trySuccess(Done)
+          // now release all cancelled tasks in phase
+          registeredPhases
+            .merge(phaseName, StrictPhaseDefinition.empty, (previous, incoming) => previous.merge(incoming))
+          if (cancelled && log.isDebugEnabled) {
+            log.debug("Successfully cancelled CoordinatedShutdown task [{}] from phase [{}].", name, phaseName)
+          }
+          cancelled
+        }
+
+        override def isCancelled: Boolean = p.isCompleted
+      }
+    }
+
+    private case class StrictPhaseDefinition(tasks: Set[TaskDefinition]) extends PhaseDefinition {
+      // This is a case class so that the update methods on ConcurrentHashMap can correctly deal with equality
+      override val size: Int = tasks.size
+
+      override def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done] = {
+        Future.sequence(tasks.map(_.run(recoverEnabled))).map(_ => Done)(ExecutionContexts.sameThreadExecutionContext)
+      }
+
+      // This method may be run multiple times during the compare-and-set loop of ConcurrentHashMap, so it must be side-effect-free
+      def merge(other: StrictPhaseDefinition): StrictPhaseDefinition = {
+        val nextTasks = (tasks ++ other.tasks).filterNot(_.isCancelled)
+        copy(tasks = nextTasks)
+      }
+    }
+
+    private object StrictPhaseDefinition {
+      def single(taskDefinition: TaskDefinition): StrictPhaseDefinition = {
+        StrictPhaseDefinition(Set(taskDefinition))
+      }
+      val empty: StrictPhaseDefinition = StrictPhaseDefinition(Set.empty)
+    }
+
+    def get(phaseName: String): Option[PhaseDefinition] = Option(registeredPhases.get(phaseName))
+
+    def totalDuration(): FiniteDuration = {
+      import akka.util.ccompat.JavaConverters._
+      registeredPhases.keySet.asScala.foldLeft(Duration.Zero) {
+        case (acc, phase) =>
+          acc + timeout(phase)
+      }
+    }
+
+    def register(phaseName: String)(task: () => Future[Done], name: String): Cancellable = {
```

Comment:
Make it a single parameter list.

Reply:
Can do. Is this a project-wide style convention (I didn't notice it in any of the docs) or a more personal preference?

Reply:
Project wide but not explicitly mentioned anywhere in docs, we only do several param lists when it is needed to help the type inference in Scala (and even then selectively), and a few cases when there is a last lambda giving a nice `method(a, b) { p => ... }` syntax, like in some of the typed actor factory APIs.